MRG: Calculate all gather stats in rust; use for rocksdb gather

bluegenes opened the pull request:
This PR adds a `calculate_gather_stats` function (and criterion benchmark) to calculate all gather statistics directly in rust. It makes use of this function in `disk_revindex.rs`, meaning that gather on a rocksdb database will now produce the full `GatherResult`. 

Note that this PR does not change the python layer to use these stats. This is just a step on the way to that :).

## New functions:

- `Sketch::MinHash::sum_abunds()` - sum abundances in a MinHash
- `Sketch::MinHash::n_unique_kmers()` - len(self) * self.scaled.
- `Sketch::inflate(abunds_from)` - inflate abundances in self using abund values from `abunds_from` MinHash
- `Sketch::inflated_abundances(abunds_from)` - same process as `inflate`, but just return the abundance vector and sum of abundances.
- `ani_utils::ani_from_containment` - a streamlined function to get ANI from containment. I suspect a lot of the time added for ANI calculation (ref: testing over in branchwater) is the way we calculate it in python (including calculating a total number of k-mers). This excludes anything not really needed for basic ANI without confidence intervals.
- `ani_utils::ani_from_containment_ci` - get high and low CI ANI from containment
- `index::calculate_gather_stats` - calculate all gather statistics that can be separated from the gather iteration and return the full `GatherResult`

The "expensive" calculations:
- abundance-weighted values require re-calculating intersections and/or manipulating the abundance vector
  > Thankfully I think the new `inflate`/`inflated_abundances` code is pretty efficient? It uses the `itertools` `merge_join_by` strategy from https://github.com/sourmash-bio/sourmash/compare/lirber/itertools_merge. Thanks @luizirber :)
- abundance `median` and `stddev` currently require cloning abunds, not sure how big of an issue that is.
- ANI with confidence intervals

## Punted Issues
- #3020: calculate all gather stats for all other database types. start with `linear.rs` since it uses `GatherResult` struct already.
- #3021 try moving `calculate_gather_stats` outside of the `gather` functions to enable postprocessing in parallel.
- #3022 : write ffi for the new functions?
  - `sketch::minhash::sum_abunds`
  - `sketch::minhash::n_unique_kmers`,
  - `sketch::minhash::inflate`
  - `sketch::minhash::inflated_abundances`
  - `ani_utils::ani_from_containment`
  - `ani_utils::ani_from_containment_ci`
  - `ani_utils::prob_nothing_in_common`

ref https://github.com/sourmash-bio/sourmash_plugin_branchwater/issues/187.

bluegenes committed:
init gather col additions

bluegenes committed:
load sig directly from record, which we already have w/ iter, iter_mut

bluegenes committed:
add potential new fns

bluegenes committed:
add manifest scaled select, tests:collection loading+downsampling

bluegenes committed:
clean up imports

bluegenes committed:
test get_sketch

bluegenes committed:
manifest from pathlist

bluegenes committed:
test minhash, get_sketch failures

bluegenes committed:
add pathbuf failure tests

bluegenes committed:
add test for trying to build manifest from bad sig path

bluegenes committed:
handle scaled selection for num sigs

bluegenes committed:
Merge branch 'latest' into core-upds-for-branchwater-plugin

bluegenes committed:
fix num selection

bluegenes committed:
Merge branch 'core-upds-for-branchwater-plugin' of github.com:sourmash-bio/sourmash into core-upds-for-branchwater-plugin

bluegenes committed:
Merge branch 'latest' into core-upds-for-branchwater-plugin

bluegenes committed:
back to concise version

bluegenes committed:
Merge branch 'core-upds-for-branchwater-plugin' into upd-gatherresult

bluegenes committed:
add inflate

bluegenes committed:
Merge branch 'latest' into upd-gatherresult

bluegenes committed:
test inflate

bluegenes committed:
Merge branch 'latest' into upd-gatherresult

bluegenes committed:
add inflated_abunds

bluegenes committed:
upd cols; add test

bluegenes committed:
fix issues

bluegenes committed:
upd rust test

bluegenes commented:
Ok, this has a lot of the basic structure in place -  I could use an initial review.

questions primarily for @luizirber:

1. Ideally we could parallelize these stats calculations separately from the actual gather operation. However, I do notice that we have to pass around a few more things -- namely minhashes for the original query and the gather-subtracted query. I'd appreciate some thoughts/direction on doing this properly. [minor update: I now see that `linear` gather does things differently, subtracting hashes from the matches instead. Could make two versions, perhaps, if this structure is useful...]
2. You were originally using `KmerMinHashBTree` for the gather-subtracted query (commented out b/c it wasn't previously used). Should I switch to that? Or is there a better way to get the subtracted query so we can use it for abundance calculations?
3. How were you thinking of doing median/stddev abundance calculations? For now I'll try out `streaming-stats` and `statistical` and use the faster one.

questions primarily for @ctb:
1. Can you sanity check `calculate_gather_stats` please?
2. We could actually return all columns used by python `GatherResult` from rust, preventing any intersections/recalculations in the python layer (and ultimately sidestepping the python GatherResult object entirely). 

I've currently calculated everything except:
  - `sum_weighted_found`
  - `potential_false_negative`
   
... and the following no-calculation (tracking) columns that come directly from the query. We could add these separately while writing the gather csv instead of 'calculating' with gather stats. Also relevant b/c query_filename and query_name are not actually available from `gather` - we only have query minhash, not sig.
- `query_filename`
- `query_name`
- `query_md5`
- `query_bp`
- `ksize`
- `moltype`
- `scaled`

bluegenes committed:
apply clippy suggestions

bluegenes committed:
addl rust tests

bluegenes committed:
roll back chrono

bluegenes committed:
make all gatherresult fields public for now

bluegenes committed:
try downsampling

bluegenes committed:
handle downsampling in main gather

bluegenes committed:
Merge branch 'latest' into upd-gatherresult

bluegenes committed:
Merge branch 'latest' into upd-gatherresult

bluegenes committed:
add streamlined ani calc

bluegenes committed:
init bench framework for gather_stats

bluegenes committed:
direct fn instead of through struct; get benchmark working

bluegenes committed:
fmt

bluegenes commented:
hi @ctb, @luizirber - an update:

I found using the `FastGatherResult` frustrating, so I removed it for now and am just passing arguments directly into `calculate_gather_stats`. I added some boolean values for whether or not to calculate abundance-weighted values and ANI via streamlined ANI eqn, and added a benchmark to see how the timing compares.

```
 cargo bench --bench gather                                                                                                                                                                  ✔  2689  08:10:25

   Compiling sourmash v0.12.1 (/Users/ntward/dib-lab/sourmash/src/core)
    Finished bench [optimized] target(s) in 6.22s
     Running benches/gather.rs (target/release/deps/gather-6938be9450cafe63)
Gnuplot not found, using plotters backend
gather_stats/abund0_ani0
                        time:   [2.8955 µs 2.9017 µs 2.9086 µs]
Found 13 outliers among 100 measurements (13.00%)
  9 (9.00%) high mild
  4 (4.00%) high severe
gather_stats/abund1_ani0
                        time:   [2.7720 µs 2.7784 µs 2.7850 µs]
Found 8 outliers among 100 measurements (8.00%)
  6 (6.00%) high mild
  2 (2.00%) high severe
gather_stats/abund0_ani1
                        time:   [2.9180 µs 2.9358 µs 2.9669 µs]
Found 9 outliers among 100 measurements (9.00%)
  6 (6.00%) high mild
  3 (3.00%) high severe
gather_stats/abund1_ani1
                        time:   [2.8029 µs 2.8096 µs 2.8174 µs]
Found 10 outliers among 100 measurements (10.00%)
  5 (5.00%) high mild
  5 (5.00%) high severe
```

Perhaps I should try some larger sigs to get a more realistic comparison?

ctb commented:
not used to interpreting these - you're not seeing a substantial difference?

bluegenes commented:
That's my interpretation - the difference seems quite small to me.

This implementation is certainly more optimized/less bloated than the python version for generating the full GatherResult (in addition to the benefits of being in the rust layer).

I think adding some full file benchmarks might help me gain a better intuition for the impact, but I haven't yet thought through how to set those up in a systematic way.

ctb commented:
K - given the new functions in Rust, it's hard for me to imagine that it's not going to be MUCH faster in practice - just moving the `for` loops over from Python to Rust should be a dramatic improvement.

I have various metagenomes to offer, ranging from SRR606249 (podar) to some really annoying rumen data sets. See `/home/ctbrown/scratch3/SRR606249-feb13-2024` for the podar data set, and the various `~ctbrown/scratch3/slainte-*/sketches/metag/` for various gut and rumen metagenomes.

ctb commented:
(as a side note, my experience with `calc-full-gather` https://github.com/sourmash-bio/sourmash/issues/2950 suggests that the next frontier *after* moving sum-abundances into Rust is speeding up FracMinHash comparisons, since they are just inherently ...slow. This is not a complaint, just an observation :)

bluegenes commented:
> (as a side note, my experience with calc-full-gather https://github.com/sourmash-bio/sourmash/issues/2950 suggests that the next frontier after moving sum-abundances into Rust is speeding up FracMinHash comparisons, since they are just inherently ...slow. This is not a complaint, just an observation :)

so maybe the right end point for this PR is just to add all calculations and use them in all index `gather` functions, leaving parallelization / other refactoring for another PR

ctb commented:
> so maybe the right end point for this PR is just to add all calculations and use them in all index `gather` functions, leaving parallelization / other refactoring for another PR

strong agree!

mr-eyes committed:
[MRG] Adding cANI Rust implementation to PR #2943 (#3005)

# Rust implementation for ANI estimation

## Usage example:

```Rust

use crate::c_ani::containment_to_distance;

fn main() {
    let containment = 0.1;
    let ksize = 31;
    let scaled = 1000; // u64
    let n_unique_kmers = 1000;
    let sequence_len_bp = n_unique_kmers * scaled;
    let confidence = Some(0.95);
    let estimate_ci = Some(true);
    let prob_threshold = Some(1e-3);

    let result = containment_to_distance(
        containment,
        ksize,
        scaled,
        Some(n_unique_kmers),
        Some(sequence_len_bp),
        confidence,
        estimate_ci,
        prob_threshold,
    );


    match result {
        Ok(ci_result) => {
            let ani_result = 1.0 - ci_result.point_estimate;
            println!("ANI: {}", ani_result);
        }

        Err(e) => println!("Error occurred: {:?}", e),
    }
}
```

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Tessa Pierce Ward <bluegenes@users.noreply.github.com>

bluegenes committed:
update cargo lock

bluegenes committed:
add utils for ani with confidence intervals

bluegenes committed:
add ani ci vals to GatherResult

bluegenes committed:
properly use/update ci values

bluegenes commented:
In contrast to standard ANI, calculating confidence intervals adds significant time (`<3 µs`  => `>7 µs`) :).

```
cargo bench --bench gather                                                                                                                                                         101 ↵  2855  14:35:03

   Compiling sourmash v0.12.1 (/Users/ntward/dib-lab/sourmash/src/core)
    Finished bench [optimized] target(s) in 6.25s
     Running benches/gather.rs (target/release/deps/gather-2811eef1e2961549)
Gnuplot not found, using plotters backend
gather_stats/abund0_ani_ci0
                        time:   [2.8582 µs 2.8711 µs 2.8853 µs]
                        change: [-2.9445% -2.0587% -1.2453%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) high mild
  2 (2.00%) high severe
gather_stats/abund1_ani_ci0
                        time:   [2.7471 µs 2.7649 µs 2.7833 µs]
                        change: [+0.2129% +1.0007% +1.8708%] (p = 0.02 < 0.05)
                        Change within noise threshold.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe
gather_stats/abund0_ani_ci1
                        time:   [7.7871 µs 7.8560 µs 7.9398 µs]
                        change: [+168.23% +176.26% +184.35%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 14 outliers among 100 measurements (14.00%)
  3 (3.00%) high mild
  11 (11.00%) high severe
gather_stats/abund1_ani_ci1
                        time:   [7.7260 µs 7.7659 µs 7.8116 µs]
                        change: [+174.13% +176.24% +178.89%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 8 outliers among 100 measurements (8.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild
  4 (4.00%) high severe
```

bluegenes committed:
streamline further

bluegenes committed:
clippy allow too many args for now

bluegenes committed:
add sum_weighted_found

bluegenes committed:
Merge branch 'latest' into upd-gatherresult

luizirber commented on src/core/benches/gather.rs:
```diff
@@ -0,0 +1,68 @@
+#[macro_use]
+extern crate criterion;
```
Drop this

```suggestion
```

luizirber commented on src/core/benches/gather.rs:
```diff
@@ -0,0 +1,68 @@
+use sourmash::signature::Signature;
+use sourmash::sketch::Sketch;
+
+use criterion::{black_box, Criterion};
```
... and add these here:
```suggestion
use criterion::{black_box, criterion_group, criterion_main, Criterion};
```

luizirber commented on src/core/benches/gather.rs:
```diff
@@ -0,0 +1,68 @@
+                    black_box(200),
+                    black_box(calc_abund_stats),
+                    black_box(calc_ani_ci),
+                );
```
`Result` is assigned but not checked here, so not necessarily doing any calculation. Replace with

```suggestion
                calculate_gather_stats(
                    black_box(&orig_query),
                    black_box(&query),
                    black_box(match_sig.clone()),
                    black_box(&match_mh),
                    black_box(42),  // Example match_size
                    black_box(100), // Example remaining_hashes
                    black_box(1),   // Example gather_result_rank
                    black_box(200),
                    black_box(calc_abund_stats),
                    black_box(calc_ani_ci),
                ).expect("error calculating gather stats");
```

(now it errors out :sweat_smile:
```
Benchmarking gather_stats/abund1_ani_ci0: Warming up for 3.0000 sthread 'main' panicked at src/core/benches/gather.rs:58:18:
error calculating gather: NeedsAbundanceTracking
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
)

luizirber commented on src/core/src/ani_utils.rs:
```diff
@@ -0,0 +1,137 @@
+
+use crate::Error;
+use roots::{find_root_brent, SimpleConvergency};
+use statrs::distribution::{ContinuousCDF, Normal};
```
```suggestion
use roots::{find_root_brent, SimpleConvergency};
use statrs::distribution::{ContinuousCDF, Normal};

use crate::Error;
```

luizirber commented on src/core/src/index/mod.rs:
```diff
@@ -11,18 +11,24 @@ pub mod revindex;
 use getset::{CopyGetters, Getters, Setters};
 
 use serde::{Deserialize, Serialize};
 use typed_builder::TypedBuilder;
```
```suggestion
use std::path::Path;

use getset::{CopyGetters, Getters, Setters};
use serde::{Deserialize, Serialize};
use stats::{median, stddev};
use typed_builder::TypedBuilder;
```

luizirber commented on src/core/src/index/mod.rs:
```diff
@@ -145,3 +200,232 @@ where
+        )
+        .unwrap();
+        // first, print all results
+        eprintln!("result: {:?}", result);
```
```suggestion
        dbg!(result);
```

luizirber commented:
Something that `SigStore` supports is `.unload()`[^2], which would allow grabbing a `SigStore` for a match, calculating stats, unloading it, and later re-loading the data from storage if needed [^1], like for `--save-matches`. Nodes/leaves in the Python SBT have a similar method, and `SigStore` is pretty much the same idea as `SigLeaf` over there, a pointer to a sig + storage that avoids keeping them in memory.

So instead of passing `match_sig` and `match_mh`, could pass a `SigStore` for `match_sig` instead, unload it after calculations, and that would have at most one sig loaded in mem (if calculating gather sequentially)

[^1]: I **think** this is done automatically, but would need to check
[^2]: if backed by a storage, doesn't do anything if in-memory

> 1. Ideally we could parallelize these stats calculations separately from the actual gather operation. However, I do notice that we have to pass around a few more things -- namely minhashes for the original query and the gather-subtracted query. I'd appreciate some thoughts/direction on doing this properly. [minor update: I now see that `linear` gather does things differently, subtracting hashes from the matches instead. Could make two versions, perhaps, if this structure is useful...]

Yeah, for the way that `linear` gather works, this came up during greyhound, more background:
- https://github.com/sourmash-bio/sourmash/issues/1226#issuecomment-720815108
- https://github.com/sourmash-bio/sourmash/issues/1957#issuecomment-1100711337 <- this one was relevant for in-memory revindex, but not so useful for on-disk revindex?
So the idea would be to create the rank (order of matches in gather), and later do the calculations in parallel. But this is easier to fix if the calculations are already in place, so we can check the refactor is correct =]
 
>     2. You were originally using `KmerMinHashBTree` for the gather-subtracted query (commented out b/c it wasn't previously used). Should I switch to that? Or is there a better way to get the subtracted query so we can use it for abundance calculations?

anything editing a minhash benefits from the `BTree` version, because the regular one is based on a `Vec` and need to potentially shuffle/reallocate chunks of hashes...

>     3. How were you thinking of doing median/stddev abundance calculations? For now I'll try out `streaming-stats` and `statistical` and use the faster one.

I usually choose deps based on:
- how many other deps does it bring along?
- does it work on wasm?
- is the API easy to use?

But no strong feelings on any specific crate, would need to check them out

bluegenes committed:
Apply suggestions from code review

Co-authored-by: Luiz Irber <luizirber@users.noreply.github.com>

bluegenes committed:
actually use abund sigs

bluegenes committed:
use KmerMinHashBTree for query

bluegenes committed:
use SigStore instead of passing in match_mh

bluegenes committed:
clean up

bluegenes commented:
Ok, the benchmarks are actually executing now (thanks @luizirber!):

```
gather_stats/abund0_ani_ci0
                        time:   [20.185 µs 20.594 µs 21.070 µs]
                        change: [-3.2668% -1.7352% +0.0139%] (p = 0.04 < 0.05)
                        Change within noise threshold.
Found 9 outliers among 100 measurements (9.00%)
  4 (4.00%) high mild
  5 (5.00%) high severe
gather_stats/abund1_ani_ci0
                        time:   [88.649 µs 90.374 µs 92.082 µs]
                        change: [-2.2100% +1.0175% +4.3261%] (p = 0.54 > 0.05)
                        No change in performance detected.
gather_stats/abund0_ani_ci1
                        time:   [27.328 µs 28.451 µs 29.648 µs]
                        change: [-11.250% -7.4807% -3.4200%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  6 (6.00%) high mild
  5 (5.00%) high severe
gather_stats/abund1_ani_ci1
                        time:   [85.135 µs 87.003 µs 89.019 µs]
                        change: [-0.7703% +2.3976% +5.5857%] (p = 0.14 > 0.05)
                        No change in performance detected.
Found 19 outliers among 100 measurements (19.00%)
  6 (6.00%) high mild
  13 (13.00%) high severe
```

From this, it looks like confidence intervals aren't too bad, but abundance does take quite a bit longer.

bluegenes committed:
Merge branch 'latest' into upd-gatherresult

bluegenes committed:
update test

bluegenes commented:
thanks for your review @luizirber!

> So instead of passing `match_sig` and `match_mh`, could pass a `SigStore` for `match_sig` instead, unload it after calculations, and that would have at most one sig loaded in mem (if calculating gather sequentially)

Good point! The downside (+ reason I passed it originally) is that we also need `match_mh` for the query subtraction, so now we end up loading it both in gather and in calculate_gather_stats. But I think if we parallelize `calculate_gather_stats`, passing just `SigStore` instead should help with memory.

Do I need to do any explicit `unload` for SigStore, or does it all happen internally?

Do we need an issue to add `inflate` and `inflated_abundances` for `KmerMinHashBTree`, or is it ok that I just added them to `KmerMinHash`?

Thanks for the info re: crate selection! I ended up using `streaming-stats`, partly because it doesn't have a lot of deps, is designed for large datasets, and provides functions for `median` and `stddev` directly, rather than requiring additional calculations. The only issue is I need to `clone` abund vectors at the moment - I'll make an issue to revisit this.

bluegenes committed:
Merge branch 'latest' into upd-gatherresult

bluegenes committed:
better remaining_bp

bluegenes committed:
Merge branch 'upd-gatherresult' of github.com:sourmash-bio/sourmash into upd-gatherresult

bluegenes committed:
add tests for ani eqns

bluegenes committed:
upd test and fix a few calcs

bluegenes committed:
upd comment

bluegenes committed:
Merge branch 'latest' into upd-gatherresult

bluegenes committed:
harmonize names, fix weighted sums

bluegenes committed:
Merge branch 'upd-gatherresult' of github.com:sourmash-bio/sourmash into upd-gatherresult

bluegenes committed:
upd GatherResult: only write ani_ci cols if they exist

bluegenes commented:
ready for review! @sourmash-bio/devs

bluegenes committed:
clippy fix

ctb commented:
> error: package `clap v4.5.0` cannot be built because it requires rustc 1.74 or newer, while the currently active rustc version is 1.65.0

ctb commented:
At a high level, this looks fine to me! I'm not really able to evaluate the Rust code ;). It's probably worth noting in the PR description that this doesn't actually change the Python calculations at all - this just makes the necessary functions in Rust.

There are a number of tests with missing coverage, is that intentional or a bug in coverage calculations?

bluegenes committed:
test var_n_mutated

bluegenes committed:
test prob nothing in common

bluegenes commented:
> There are a number of tests with missing coverage, is that intentional or a bug in coverage calculations?

Lost track of a few fns that needed testing. Added remaining tests.

There are a few codecov lines that I don't fully understand -- hoping they get updated as covered.

luizirber committed:
Roll back cargo lock to work on 1.65

luizirber commented on src/core/benches/gather.rs:
```diff
@@ -0,0 +1,62 @@
+    filename.push("../../tests/test-data/track_abund/47.fa.sig");
+    let file = File::open(filename).unwrap();
+    let reader = BufReader::new(file);
+    let mut sigs: Vec<Signature> = serde_json::from_reader(reader).expect("Loading error");
+    let orig_query = if let Sketch::MinHash(mh) = &sigs.swap_remove(0).sketches()[0] {
+        mh.clone()
+    } else {
+        unimplemented!()
+    };
```
Let's put the new `.minhash()` method to use =]

```suggestion
    let sig = Signature::from_reader(reader).expect("Loading error").swap_remove(0);                                                                                                  
    let orig_query = sig.minhash().unwrap();
```

luizirber commented on src/core/Cargo.toml:
```diff
@@ -49,11 +50,14 @@ once_cell = "1.18.0"
+roots = "0.0.8"
 serde = { version = "1.0.196", features = ["derive"] }
 serde_json = "1.0.113"
+statrs = "0.16.0"
+streaming-stats = "0.2.3"
```
I didn't look more deeply into them, but I guess we can't consolidate into one?

bluegenes replied:
`statrs` does implement a `statistics::Median` trait and `statistics::VarianceN` trait, but I don't know how to use them.

bluegenes replied:
upon digging a little further, `statrs` provides distributions (e.g. `ContinuousCDF`, `Normal`) that we need for ANI confidence intervals, while `streaming-stats` is used for `median` and `stddev` directly from a vector of abundances. I think we need both (or a different stats library entirely).

luizirber commented on src/core/src/index/mod.rs:
```diff
@@ -14,15 +14,21 @@ pub mod search;
+use crate::storage::SigStore;
 use crate::Result;
 
+// Todo: evaluate get vs get_copy for added fields
```
```suggestion
```

luizirber committed:
Replace Signature for SigStore in GatherResult

- change signature for remove_many, takes an iterator now
- implement PartialEq for SigStore

luizirber committed:
clippy fixes

luizirber reviewed (approved):
This is so cool, thanks Tessa!

luizirber commented:
I piggybacked some commits replacing `Signature` with `SigStore` in `GatherResults`, just to find out that the `.unload()` method on `SigStore` is not implemented yet, oops! But now things are in place for it to work =]

I also changed `.remove_many()` on `KmerMinHash`  to take an `Iterator`, this way we don't need to allocate a new vec to remove hashes...

The last commit is for some clippy fixes, I suspect CI is not running clippy with all features enabled, and some got away

luizirber committed:
update C headers

bluegenes committed:
Apply suggestions from code review

Co-authored-by: Luiz Irber <luizirber@users.noreply.github.com>

bluegenes committed:
add mh test; rustfmt

bluegenes committed:
Merge branch 'latest' into upd-gatherresult
